File: backend/src/mandate-publish.ts
// backend/src/mandate-publish.ts
//
// Implementa Encargo_Implementacion_Publicacion_Mandate_Bootstrap_v1_0.md §2.1.
//
// Cierra el tramo entre "mandate firmado en disco" (nucleus mandate build) y "fila en
// Backend que permite que resolveMandateDelivery (mandate-delivery.ts, sin cambios) la
// encuentre". No reabre resolveMandateDelivery ni el schema de Marketplace ya existente
// (mandates / mandate_versions / organization_bootstrap_mandates, migraciones
// 0000_initial.sql / 0003_organization_bootstrap_mandates.sql, bucket R2 MANDATES) — sólo
// agrega el escritor que faltaba.
//
// Alcance de esta ronda (§1 del encargo, no se resuelve acá — señalado en el cierre):
// sólo el camino feliz de "primer mandate de una organización nueva" (mandateType
// genesis). Metadata mínima: description default '' (columna NOT NULL), pillar/origin_type
// quedan NULL (no se exponen en el body todavía).

import { sha256HexBytes } from './authority/canonical';

export class PublishError extends Error {
  constructor(public status: 400 | 409 | 500, message: string) { super(message); }
}

export interface PublishMandateBody {
  slug: string;
  version: string;
  description?: string;
  visibility: 'private' | 'public';
  mandate_base64: string;
  sha256: string;
  bootstrap: boolean;
}

export interface PublishMandateResult {
  mandate_id: string;
  mandate_version_id: string;
  bootstrap_assigned: boolean;
}

const HEX_64 = /^[a-fA-F0-9]{64}$/;

function base64ToBytes(base64: string): Uint8Array {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function validateBody(body: PublishMandateBody | null): asserts body is PublishMandateBody {
  if (!body) throw new PublishError(400, 'invalid_body');
  if (!body.slug || typeof body.slug !== 'string') throw new PublishError(400, 'missing_slug');
  if (!body.version || typeof body.version !== 'string') throw new PublishError(400, 'missing_version');
  if (body.visibility !== 'private' && body.visibility !== 'public') throw new PublishError(400, 'invalid_visibility');
  if (!body.mandate_base64 || typeof body.mandate_base64 !== 'string') throw new PublishError(400, 'missing_mandate_base64');
  if (!body.sha256 || !HEX_64.test(body.sha256)) throw new PublishError(400, 'invalid_sha256');
  if (typeof body.bootstrap !== 'boolean') throw new PublishError(400, 'missing_bootstrap');
}

/**
 * Publica un mandate firmado: sube el artefacto a R2, registra/actualiza la fila
 * `mandates` y su `mandate_versions`, y opcionalmente lo asigna como bootstrap de la
 * organización. Espejo, del lado de escritura, de lo que resolveMandateDelivery
 * (mandate-delivery.ts) ya hace del lado de lectura — misma tabla, mismo bucket, mismo
 * criterio de integridad (sha256 recalculado, nunca confiado a ciegas).
 *
 * Orden de pasos: sigue literalmente §2.1 del encargo, incluyendo que el paso 2 (upsert
 * de `mandates`) corre ANTES del chequeo de idempotencia/conflicto del paso 3 — un intento
 * con `sha256` distinto para una versión ya publicada devuelve 409 en el paso 3, pero
 * `mandates.latest_version` ya quedó actualizado en el paso 2. Ese orden es el que pide el
 * diseño cerrado, no una elección de esta implementación.
 */
export async function publishMandate(env: Env, organizationId: string, rawBody: PublishMandateBody | null, now = new Date()): Promise<PublishMandateResult> {
  validateBody(rawBody);
  const body = rawBody;

  // Paso 1: decodificar y recalcular sha256 — rechazo antes de tocar la base.
  let bytes: Uint8Array;
  try {
    bytes = base64ToBytes(body.mandate_base64);
  } catch {
    throw new PublishError(400, 'invalid_mandate_base64');
  }
  const digest = await sha256HexBytes(bytes);
  if (digest !== body.sha256.toLowerCase()) throw new PublishError(400, 'sha256_mismatch');

  const publishedAt = now.getTime();

  // Paso 2: mandates — buscar por (origin_org_id, slug); crear o actualizar latest_version.
  const existingMandate = await env.DB.prepare('SELECT id FROM mandates WHERE origin_org_id = ? AND slug = ?')
    .bind(organizationId, body.slug).first<{ id: string }>();
  const mandateId = existingMandate?.id ?? crypto.randomUUID();
  if (existingMandate) {
    await env.DB.prepare('UPDATE mandates SET latest_version = ? WHERE id = ?')
      .bind(body.version, mandateId).run();
  } else {
    await env.DB.prepare(
      `INSERT INTO mandates (id, origin_org_id, slug, description, visibility, latest_version, pillar, origin_type, created_at)
       VALUES (?, ?, ?, ?, ?, ?, NULL, NULL, ?)`,
    ).bind(mandateId, organizationId, body.slug, body.description ?? '', body.visibility, body.version, publishedAt).run();
  }

  // Paso 3: mandate_versions — idempotencia por (mandate_id, version).
  const existingVersion = await env.DB.prepare('SELECT id, sha256 FROM mandate_versions WHERE mandate_id = ? AND version = ?')
    .bind(mandateId, body.version).first<{ id: string; sha256: string }>();

  let mandateVersionId: string;
  if (existingVersion) {
    if (existingVersion.sha256.toLowerCase() !== digest) throw new PublishError(409, 'version_conflict');
    // Retry idempotente: mismo sha256, no se vuelve a subir a R2 ni a insertar.
    mandateVersionId = existingVersion.id;
  } else {
    mandateVersionId = crypto.randomUUID();
    const r2Key = `mandates/${organizationId}/${body.slug}/${body.version}.json`;
    // Paso 4: subir a R2 — sólo acá, porque el paso 3 no encontró ya la versión.
    await env.MANDATES.put(r2Key, bytes);
    await env.DB.prepare(
      'INSERT INTO mandate_versions (id, mandate_id, version, r2_key, sha256, published_at) VALUES (?, ?, ?, ?, ?, ?)',
    ).bind(mandateVersionId, mandateId, body.version, r2Key, digest, publishedAt).run();
  }

  // Paso 5: bootstrap — upsert de organization_bootstrap_mandates (PK organization_id).
  // assigned_at sólo si la fila es nueva; updated_at siempre.
  let bootstrapAssigned = false;
  if (body.bootstrap) {
    const existingBootstrap = await env.DB.prepare('SELECT organization_id FROM organization_bootstrap_mandates WHERE organization_id = ?')
      .bind(organizationId).first<{ organization_id: string }>();
    if (existingBootstrap) {
      await env.DB.prepare('UPDATE organization_bootstrap_mandates SET mandate_version_id = ?, updated_at = ? WHERE organization_id = ?')
        .bind(mandateVersionId, publishedAt, organizationId).run();
    } else {
      await env.DB.prepare(
        'INSERT INTO organization_bootstrap_mandates (organization_id, mandate_version_id, assigned_at, updated_at) VALUES (?, ?, ?, ?)',
      ).bind(organizationId, mandateVersionId, publishedAt, publishedAt).run();
    }
    bootstrapAssigned = true;
  }

  return { mandate_id: mandateId, mandate_version_id: mandateVersionId, bootstrap_assigned: bootstrapAssigned };
}

File: installer/nucleus/internal/orchestration/commands/mandate_publish.go
// internal/orchestration/commands/mandate_publish.go
//
// Implementa Encargo_Implementacion_Publicacion_Mandate_Bootstrap_v1_0.md §2.2.
//
// Cierra el tramo entre "mandate firmado en disco" (mandate.json escrito por
// SignMandateActivity, ver mandate_genesis_sign_activity.go — sin cambios) y "fila en
// Backend" (POST /v1/mandate/publish, ver mandate-publish.ts) — publica el artefacto ya
// firmado hacia el Marketplace de Backend y, opcionalmente, lo asigna como bootstrap de
// la organización activa.
//
// Alcance de esta ronda (§1 del encargo): sólo el camino feliz de "primer mandate de una
// organización nueva" (mandateType genesis), sin body-binding en la firma S2S (reusa
// BLOOM-INSTALLATION-AUTH-v1, mismo criterio que mandatedelivery.Client.Receive), y
// metadata mínima (description default cadena vacía, pillar/origin_type no se exponen).
//
// GAP DE DISEÑO señalado (no resuelto acá, ver cierre): el encargo lista los flags de
// este comando en §2.2 (--id, --authority-url, --slug, --visibility, --description,
// --bootstrap) pero NO incluye --version — y sin embargo el contrato de Backend
// (PublishMandateBody en mandate-publish.ts, §2.1) exige `version` como campo
// obligatorio, y mandate.json (MandateJSON en mandate_genesis_sign_activity.go) no tiene
// ningún campo de versión: sólo mandateId/mandateType/project/status/signedAt/operational.
// Se agrega acá un flag --version con default "1.0.0" (primera publicación de un mandate
// genesis recién nacido) para poder cumplir el contrato de Backend — no es parte de la
// tabla de archivos/flags que cerró el encargo, así que se señala explícitamente en vez
// de asumirlo en silencio.
//
// EXTRACCIÓN DE FIRMA (§2.2 paso 4): el encargo pide extraer la construcción de firma
// BLOOM-INSTALLATION-AUTH-v1 a una función compartida en internal/authority si no existe
// ya una reutilizable. Se confirmó (grep) que NO existe: authority.SyncClient.signedRequest
// (sync.go) es un método privado atado a Binding/SyncClient, y mandatedelivery.Client.Receive
// (client.go) la repite inline — ninguna de las dos está exportada de forma reusable desde
// otro paquete. Extraerla a internal/authority requeriría tocar un archivo fuera de la
// tabla exacta de §3 de este encargo ("no agregues ni saques nada de ahí sin avisar"), así
// que se prioriza la tabla de archivos y esta construcción queda inline acá (privada a
// este archivo, no en mandatedelivery ni sync.go — cero cambios ahí, igual que pide §2.3).
// Señalado en el cierre para que se decida aparte si vale la pena esa extracción.
package commands

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"nucleus/internal/authority"
	"nucleus/internal/core"
	"nucleus/internal/supervisor"

	"github.com/spf13/cobra"
)

// sha256HexOf recalcula el sha256 del mandate leído del disco — nunca se confía en un
// digest ya guardado en otro lado, mismo criterio de integridad que
// resolveMandateDelivery (Backend) usa del lado de lectura.
func sha256HexOf(raw []byte) string {
	sum := sha256.Sum256(raw)
	return hex.EncodeToString(sum[:])
}

const mandatePublishInstallationAuthDomain = "BLOOM-INSTALLATION-AUTH-v1"

// mandateJSONStatus es la forma mínima de mandate.json que este comando necesita leer —
// sólo status, para el chequeo de precondición. No reimplementa MandateJSON (definida en
// mandate_genesis_sign_activity.go, package orchestration/activities, no importable acá
// sin crear un acoplamiento nuevo entre commands y activities que el encargo no pidió) —
// decodifica sólo el campo que necesita, igual que ya hace mandateStatusSubcommand más
// arriba en este mismo paquete (mandate.go) al mostrar mandate.json crudo.
type mandateJSONStatus struct {
	Status string `json:"status"`
}

// publishMandateRequestBody espeja PublishMandateBody de mandate-publish.ts (Backend,
// §2.1 del encargo) — mismos nombres de campo, mismo criterio de obligatoriedad.
type publishMandateRequestBody struct {
	Slug          string `json:"slug"`
	Version       string `json:"version"`
	Description   string `json:"description,omitempty"`
	Visibility    string `json:"visibility"`
	MandateBase64 string `json:"mandate_base64"`
	SHA256        string `json:"sha256"`
	Bootstrap     bool   `json:"bootstrap"`
}

type publishMandateResponseBody struct {
	MandateID         string `json:"mandate_id"`
	MandateVersionID  string `json:"mandate_version_id"`
	BootstrapAssigned bool   `json:"bootstrap_assigned"`
}

// PublishMandateResult es lo que imprime el comando (--json o interactivo).
type PublishMandateResult struct {
	Success           bool   `json:"success"`
	MandateID         string `json:"mandate_id"`
	MandateVersionID  string `json:"mandate_version_id,omitempty"`
	BootstrapAssigned bool   `json:"bootstrap_assigned,omitempty"`
	Error             string `json:"error,omitempty"`
}

var mandatePublishHTTPClient *http.Client
var mandatePublishNow = func() time.Time { return time.Now().UTC() }

func createPublishMandateSubcommand(c *core.Core) *cobra.Command {
	var mandateID, authorityURL, slug, visibility, description, version string
	var bootstrap bool

	cmd := &cobra.Command{
		Use:   "publish",
		Short: "Publica un mandate firmado (mandate.json) hacia Backend y, opcionalmente, lo asigna como bootstrap de la organización",
		Long: `Sube el mandate ya firmado en {mandatesRoot}/{mandateID}/mandate.json a Backend
(POST /v1/mandate/publish) — recalcula su sha256, firma el request S2S con la identidad de
instalación de esta org activa (mismo dominio BLOOM-INSTALLATION-AUTH-v1 que 'nucleus
authority sync'), y publica. Con --bootstrap=true (default) esa versión queda asignada como
el mandate de bootstrap de la organización: una instalación nueva que llame a
GET /v1/mandate/bootstrap la recibe.

Requiere que 'nucleus auth link' (o 'nucleus authority sync') ya haya corrido antes — la
instalación necesita estar registrada contra Authority.`,
		Args: cobra.NoArgs,
		Annotations: map[string]string{
			"category": "MANDATES",
			"json_response": `{
  "success": true,
  "mandate_id": "3f9c1a2e-8b7d-4c1a-9e2f-1a2b3c4d5e6f",
  "mandate_version_id": "a1b2c3d4-...",
  "bootstrap_assigned": true
}`,
		},
		Example: `  nucleus mandate publish --id 3f9c1a2e-8b7d-4c1a-9e2f-1a2b3c4d5e6f --authority-url https://authority.bloom.dev --slug genesis-bootstrap
  nucleus mandate publish --id 3f9c1a2e-... --authority-url https://authority.bloom.dev --slug genesis-bootstrap --bootstrap=false
  nucleus --json mandate publish --id 3f9c1a2e-... --authority-url https://authority.bloom.dev --slug genesis-bootstrap`,
		PreRun: func(cmd *cobra.Command, args []string) {
			if err := requireMandateMaster(c); err != nil {
				fmt.Printf("Error: %v\n", err)
				os.Exit(1)
			}
		},

		Run: func(cmd *cobra.Command, args []string) {
			result := runPublishMandate(c, mandateID, authorityURL, slug, visibility, description, version, bootstrap)
			if c.IsJSON {
				data, _ := json.MarshalIndent(result, "", "  ")
				fmt.Println(string(data))
			} else if result.Success {
				c.Logger.Printf("[SUCCESS] ✅ Mandate publicado: %s (versión %s, bootstrap_assigned=%t)", result.MandateID, result.MandateVersionID, result.BootstrapAssigned)
			} else {
				c.Logger.Printf("[ERROR] ❌ No se pudo publicar el mandate: %s", result.Error)
			}
			if !result.Success {
				os.Exit(1)
			}
		},
	}
	cmd.Flags().StringVar(&mandateID, "id", "", "ID del mandate ya construido y firmado (requerido)")
	cmd.Flags().StringVar(&authorityURL, "authority-url", "", "URL base del backend de Authority (requerido)")
	cmd.Flags().StringVar(&slug, "slug", "", "Slug del mandate en el Marketplace (requerido)")
	cmd.Flags().StringVar(&visibility, "visibility", "private", "Visibilidad del mandate ('private' o 'public')")
	cmd.Flags().StringVar(&description, "description", "", "Descripción del mandate (opcional)")
	// Ver GAP DE DISEÑO en el comentario de cabecera del archivo: no estaba en la lista
	// de flags de §2.2, se agrega para poder cumplir el campo obligatorio `version` del
	// contrato de Backend (§2.1).
	cmd.Flags().StringVar(&version, "version", "1.0.0", "Versión publicada del mandate — GAP: no especificada por §2.2, ver comentario de cabecera del archivo")
	cmd.Flags().BoolVar(&bootstrap, "bootstrap", true, "Asigna esta versión como el mandate de bootstrap de la organización activa")
	_ = cmd.MarkFlagRequired("id")
	_ = cmd.MarkFlagRequired("authority-url")
	_ = cmd.MarkFlagRequired("slug")
	return cmd
}

// runPublishMandate hace el trabajo real; separado de Run para que sea testeable sin
// pasar por cobra. Logging estructurado vía core.InitLogger(&c.Paths, "MANDATE", c.IsJSON)
// — mismo mecanismo que 'nucleus auth link' (ver auth_link.go, corregido en
// Encargo_Correccion_Nucleus_Comando_Vinculacion_Organizacion_v2_0.md §2.2) y que ya
// cumple la Guía Maestra de Comandos y la spec de Telemetría — "MANDATE" ya es una
// categoría de primera clase en internal/core/logger.go (icono 📋 y descripción propios)
// antes de este cambio, así que no hace falta un logger ni una categoría nueva.
func runPublishMandate(c *core.Core, mandateID, authorityURL, slug, visibility, description, version string, bootstrap bool) PublishMandateResult {
	logger, err := core.InitLogger(&c.Paths, "MANDATE", c.IsJSON)
	if err != nil {
		return PublishMandateResult{Success: false, Error: fmt.Sprintf("no pude inicializar el logger de MANDATE: %v", err)}
	}
	defer logger.Close()

	if mandateID == "" || authorityURL == "" || slug == "" {
		return PublishMandateResult{Success: false, Error: "--id, --authority-url y --slug son requeridos"}
	}
	if visibility != "private" && visibility != "public" {
		return PublishMandateResult{Success: false, Error: fmt.Sprintf("--visibility inválido: %q (esperaba 'private' o 'public')", visibility)}
	}

	cfg, err := supervisor.LoadNucleusConfig()
	if err != nil {
		logger.Error("no pude leer nucleus.json: %v", err)
		return PublishMandateResult{Success: false, Error: fmt.Sprintf("no pude leer nucleus.json: %v", err)}
	}

	mandateBytes, err := readSignedMandateFile(cfg.MandatesRoot(), mandateID)
	if err != nil {
		logger.Error("%v", err)
		return PublishMandateResult{Success: false, Error: err.Error()}
	}

	active, err := core.ResolveActiveOrgContext()
	if err != nil {
		logger.Error("no pude resolver la organización activa: %v", err)
		return PublishMandateResult{Success: false, Error: fmt.Sprintf("no pude resolver la organización activa: %v", err)}
	}

	identityPath := filepath.Join(c.Paths.AppDataDir, "authority", "identity.json")
	identity, err := authority.LoadOrCreateLocalIdentity(identityPath)
	if err != nil {
		logger.Error("no pude cargar la identidad de instalación: %v", err)
		return PublishMandateResult{Success: false, Error: fmt.Sprintf("no pude cargar la identidad de instalación: %v", err)}
	}

	digest := sha256HexOf(mandateBytes)
	requestBody := publishMandateRequestBody{
		Slug:          slug,
		Version:       version,
		Description:   description,
		Visibility:    visibility,
		MandateBase64: base64.StdEncoding.EncodeToString(mandateBytes),
		SHA256:        digest,
		Bootstrap:     bootstrap,
	}

	logger.Info("Publicando mandate %s (slug=%s, versión=%s) contra %s", mandateID, slug, version, authorityURL)
	response, err := postMandatePublish(context.Background(), authorityURL, active.OrganizationID, identity, requestBody, mandatePublishNow())
	if err != nil {
		logger.Error("publicación fallida: %v", err)
		return PublishMandateResult{Success: false, Error: err.Error()}
	}

	logger.Success("mandate publicado: %s (versión %s, bootstrap_assigned=%t)", response.MandateID, response.MandateVersionID, response.BootstrapAssigned)
	return PublishMandateResult{
		Success:           true,
		MandateID:         response.MandateID,
		MandateVersionID:  response.MandateVersionID,
		BootstrapAssigned: response.BootstrapAssigned,
	}
}

// readSignedMandateFile lee {mandatesRoot}/{mandateID}/mandate.json, falla claro si no
// existe o si su status no es "signed" — sin llamar a Backend en ningún caso de error
// (§4 del encargo: "mandate no encontrado o no firmado → error claro sin llamar al
// Backend").
func readSignedMandateFile(mandatesRoot, mandateID string) ([]byte, error) {
	path := filepath.Join(mandatesRoot, mandateID, "mandate.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("mandate no encontrado: no existe %s — ¿corriste 'nucleus mandate build --id %s' hasta que termine?", path, mandateID)
		}
		return nil, fmt.Errorf("no pude leer %s: %w", path, err)
	}
	var status mandateJSONStatus
	if err := json.Unmarshal(raw, &status); err != nil {
		return nil, fmt.Errorf("%s no es JSON válido: %w", path, err)
	}
	if status.Status != "signed" {
		return nil, fmt.Errorf("mandate %s no está firmado todavía (status=%q en %s) — esperá a que 'mandate build' termine", mandateID, status.Status, path)
	}
	return raw, nil
}

// postMandatePublish firma y envía el POST a /v1/mandate/publish. Mismo dominio, mismo
// canonical y mismos headers que mandatedelivery.Client.Receive (client.go) y
// authority.SyncClient.signedRequest (sync.go) — ver nota de "EXTRACCIÓN DE FIRMA" en la
// cabecera del archivo sobre por qué esta construcción queda inline acá en vez de
// reexportada desde internal/authority.
func postMandatePublish(ctx context.Context, authorityURL, organizationID string, identity *authority.LocalIdentity, body publishMandateRequestBody, now time.Time) (publishMandateResponseBody, error) {
	if len(identity.PrivateKey) != ed25519.PrivateKeySize {
		return publishMandateResponseBody{}, fmt.Errorf("clave privada de instalación inválida")
	}
	const path = "/v1/mandate/publish"
	url := strings.TrimRight(authorityURL, "/") + path + "?org=" + organizationID

	timestamp := now.UTC().Format(time.RFC3339Nano)
	authPayload, _ := json.Marshal(map[string]string{
		"installation_id": identity.InstallationID,
		"organization_id": organizationID,
		"method":          http.MethodPost,
		"path":            path,
		"timestamp":       timestamp,
	})
	canonical, err := authority.Canonicalize(authPayload)
	if err != nil {
		return publishMandateResponseBody{}, fmt.Errorf("no pude canonicalizar el payload de firma: %w", err)
	}
	signature := ed25519.Sign(identity.PrivateKey, append(append([]byte(mandatePublishInstallationAuthDomain), 0), canonical...))

	encodedBody, err := json.Marshal(body)
	if err != nil {
		return publishMandateResponseBody{}, fmt.Errorf("no pude serializar el body de publish: %w", err)
	}

	request, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(encodedBody))
	if err != nil {
		return publishMandateResponseBody{}, err
	}
	request.Header.Set("Content-Type", "application/json")
	request.Header.Set("X-Bloom-Installation-Id", identity.InstallationID)
	request.Header.Set("X-Bloom-Timestamp", timestamp)
	request.Header.Set("X-Bloom-Signature", base64.StdEncoding.EncodeToString(signature))

	client := mandatePublishHTTPClient
	if client == nil {
		client = &http.Client{Timeout: 30 * time.Second}
	}
	response, err := client.Do(request)
	if err != nil {
		return publishMandateResponseBody{}, fmt.Errorf("request a %s falló: %w", url, err)
	}
	defer response.Body.Close()
	raw, err := io.ReadAll(io.LimitReader(response.Body, 1<<20))
	if err != nil {
		return publishMandateResponseBody{}, fmt.Errorf("no pude leer la respuesta de %s: %w", url, err)
	}
	if response.StatusCode != http.StatusOK {
		return publishMandateResponseBody{}, fmt.Errorf("publish HTTP %d: %s", response.StatusCode, strings.TrimSpace(string(raw)))
	}
	var parsed publishMandateResponseBody
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return publishMandateResponseBody{}, fmt.Errorf("respuesta de publish no es JSON válido: %w", err)
	}
	return parsed, nil
}
